Use the PropertyAccessor component instead of our custom solution

javiereguiluz opened the pull request.

yceruto commented:
@javiereguiluz I guess this leads to improved performance too?https://github.com/symfony/symfony/pull/16838

javiereguiluz commented:
@yceruto probably. And less lines of code for us, less bugs, less maintenance, etc.  But first, let's try to make tests green for this change :smile:

ogizanagi commented on Configuration/Configurator.php:
```diff
@@ -452,21 +452,14 @@ private function getFieldFormat($fieldType)
-                $fieldConfiguration['isReadable'] = $getter || $isPublic;
-                $fieldConfiguration['isWritable'] = $setter || $isPublic;
+                $fieldConfiguration['isReadable'] = $this->accessor->isReadable($instance , $fieldName);
+                $fieldConfiguration['isWritable'] = $this->accessor->isWritable($instance, $fieldName);
```
Unfortunately, those methods were added in 2.5 only :/
We should have a BC layer temporary until 2.3 support is dropped.

javiereguiluz replied:
These are really bad news :sob:  I'll add that BC layer then.

Pierstoval replied:
Create a decorator for the PropertyAccessor, it should be the best solution.

ogizanagi commented:
I'm all in for this ! :+1:

Pierstoval commented:
:+1: , it makes me want time to go faster for 2.3 to reach end of life, to avoid all these workarounds :stuck_out_tongue:

javiereguiluz commented:
I've refactored this code completely. It now supports 2.3, but I haven't added any BC layer. The problem is divided into two parts: read a value and set a value.

Before we had an `introspectGettersAndSetters()` that set `isReadable`, `isWritable`, `getter`, `setter`, etc. for each property of each entity and for each view. This was necessary because we used a custom solution instead of the PropertyAccessor component. Now I've removed everything.

1. When reading a value, we wrap the `readValue()` method in a `try ... catch` and we display the usual `Inaccessible` label for values that cannot be read. When we drop support for 2.3, we can replace the `try .. catch` by a proper `if .. $accessor->isReadable(...)`
2. When writing a value (this only happens in the controller), we don't check if `isWritable` and we don't wrap it in a `try .. catch` That's because in this case we wan't to fail hard. Otherwise, if we fail silently, the user may think that the value was modified. If you try to edit a value which is not writable, the app should fail in a clear way.

This change simplifies everything and shaves hundreds of lines of code. Thoughts?

ogizanagi commented on Tests/PropertyAccessor/PropertyAccessorTest.php:
```diff
@@ -0,0 +1,136 @@
+        $accessor = new PropertyAccessor();
+        $object = new FooBarClass();
+
+        $this->assertSame($expectedResult, $accessor->isReadable($object, $property));
```
Still not available as of 2.3 ^^'

javiereguiluz replied:
Nice catch. Fixed (with a `markTestSkipped` for 2.3).

ogizanagi commented on Twig/EasyAdminTwigExtension.php:
```diff
@@ -153,22 +156,23 @@ public function renderEntityField(\Twig_Environment $twig, $view, $entityName, $
-                    $templateParameters['link_parameters'] = array('entity' => $targetEntityConfig['name'], 'action' => 'show', 'view' => $view, 'id' => $value->$targetEntityPrimaryKeyGetter());
+                    try {
+                        // if the target entity has a readable primary key, we
+                        // dsiplay a link pointing to its 'show' view
```
typo:

```diff
- dsiplay
+ display
```

ogizanagi commented:
@javiereguiluz : Those changes are fine to me, especially as it simplifies things a lot. But someone might have relied one those metadata, so this is potentially a BC break I guess.
I would have opt for a decorator of the `PropertyAccessor` however, as suggested by @Pierstoval, in order to avoid [this try/catch](https://github.com/javiereguiluz/EasyAdminBundle/pull/625/files#diff-0c199003d2370f3ea2800f894b951396R165) over a simple if, and smooth the upgrade path when we'll drop 2.3 support.
I agree about hard failing when writing values.

ogizanagi commented on Tests/PropertyAccessor/PropertyAccessorTest.php:
```diff
@@ -0,0 +1,144 @@
+        $accessor = new PropertyAccessor();
+        $object = new FooBarClass();
+
+    if (!method_exists($accessor, 'isReadable')) {
```
CS.
However, if we completely rely on the Symfony PropertyAccessor, without decorating/modifying it, do we have to test it at all ?

javiereguiluz replied:
Yes. I know we should remove this test. I maintained it because we are doing the transition and I wanted to be sure that it works as we expect.  Should I remove it this very same pull request?

ogizanagi replied:
You're right. Good idea in fact to keep it until now. As soon as the PR is ready to be merged, it can be removed :smiley:

javiereguiluz commented:
Fixed all the details and added a note in the UPGRADE file (in c7c6331)

Pierstoval commented on UPGRADE.md:
```diff
@@ -27,8 +41,8 @@ Upgrade to 1.9.2
 
-Upgrade to 1.8.0
-----------------
+Upgrade to 1.8.0 (8/November/2015)
```
I don't like having the `/` in a full date when the month is written in letters.
Either put `08/11/2015` or (better I think) put `November 8th 2015`
